domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt

```diff
@@ -0,0 +1,252 @@
+  @Inject
+  lateinit var audioPlayerController: AudioPlayerController
+
+  private lateinit var shadowMediaPlayer: ShadowMediaPlayer
+
+  private val TEST_URL = "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3"
+
+  @Before
+  fun setup() {
+    setUpTestApplicationComponent()
+    addMediaInfo()
+    shadowMediaPlayer = Shadows.shadowOf(audioPlayerController.getTestMediaPlayer())
+    shadowMediaPlayer.dataSource = DataSource.toDataSource(context , Uri.parse(TEST_URL))
+  }
+
+
+  @Test
+  fun testAudioPlayer_successfulInitialize_reportsSuccessfulInit() {
+    audioPlayerController.initializeMediaPlayer(TEST_URL)
+
+    shadowMediaPlayer.invokePreparedListener()
+
+    assertThat(shadowMediaPlayer.isPrepared).isTrue()
+    assertThat(audioPlayerController.isPlaying()).isFalse()
+  }
+
+  @Test
+  fun testAudioPlayer_play_isPlaying() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+
+    assertThat(audioPlayerController.isPlaying()).isTrue()
+    assertThat(shadowMediaPlayer.isReallyPlaying).isTrue()
+  }
+
+  @Test
+  fun testAudioPlayer_pause_notIsPlaying() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.pause()
+
+    assertThat(audioPlayerController.isPlaying()).isFalse()
+    assertThat(shadowMediaPlayer.isReallyPlaying).isFalse()
+  }
+
+  @Test
+  fun testAudioPlayer_seekTo_hasCorrectProgress() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.seekTo(500)
+
+    assertThat(shadowMediaPlayer.currentPositionRaw).isEqualTo(500)
+  }
+
+  @Test
+  fun testAudioPlayer_releaseMediaPlayer_isReleasedState() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.releaseMediaPlayer()
+
+    assertThat(shadowMediaPlayer.state).isEqualTo(ShadowMediaPlayer.State.END)
+  }
+
+  @Test
+  fun testAudioObserver_invokePrepare_capturesPreparedState() {
+    audioPlayerController.getPlayState().observeForever(mockAudioPlayerObserver)
+
+    arrangeMediaPlayer()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.isSuccess()).isTrue()
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(AudioPlayerController.PlayStatus.PREPARED)
+  }
+
+  @Test
+  fun testAudioObserver_invokeCompletion_capturesCompletedState() {
+    audioPlayerController.getPlayState().observeForever(mockAudioPlayerObserver)
+    arrangeMediaPlayer()
+
+    shadowMediaPlayer.invokeCompletionListener()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.isSuccess()).isTrue()
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(AudioPlayerController.PlayStatus.COMPLETED)
+  }
+
+  @Test
+  fun testAudioObserver_invokeChangeDataSource_capturesPendingState() {
+    audioPlayerController.getPlayState().observeForever(mockAudioPlayerObserver)
+    arrangeMediaPlayer()
+
+    audioPlayerController.changeDataSource(TEST_URL)
```

Comment:
Is it possible/also worth testing that this properly sets the data source on the media player?

Reply:
shadowMediaPlayer has sourceUri field but it's always null for some reason even if I call setDataSource on mediaPlayer, so I currently can't check if the correct source was set correctly

Reply:
Ack. That's unfortunate, but thanks for checking.